Verify Correctness of Reconstructed Commitment, Update proto files

the-real-adammork opened the pull request:
Soundtrack of this PR: [Tilman - Just A Motion Away](https://soundcloud.com/moskalus/premiere-tilman-just-a-motion-away-shall-not-fade)

### Motivation

A previous change added a way to reconstruct the commitment data locally with info already sent in the TxOut protobuf. The comparison of its crc32 checksum was missing and needs to be added.

LibMobileCoin is updated with `mobilecoin` latest master which has some changes in the `consensus_common` proto files.

### In this PR
* Wrapper function around new function that computes the crc32 from a "CompressedCommitment"
* Add matching crc32's as a pre-condition to creating a `PartialTxOut`
* Update naming of `minimumFee` to `mobMinimumFee` (which is now deprecated) 

### Future Work
* Merge in changes in the `mobilecoin` repo, and the `libmobilecoin-ios-artifacts` repo
* Migrate minimum fee code to pull from the new map structure.

the-real-adammork committed:
Add wrapper for crc32 func that goes into libmobile-coin, Add matching crc32 pre-condition to PartialTxOut init (they need to match)

the-real-adammork committed:
update to latest libmobile coin

the-real-adammork committed:
update lockfile, change proto to use now-depercated mobMinimumFee

the-real-adammork committed:
update libmobilecoin hash

the-real-adammork committed:
Update to latest libmobilecoin commit, Add unit test for commitment reconstruction and crc32 calculation. adjust function signature of some calls. Update incorrect comments

the-real-adammork committed:
make a release from latest changes, udpate change log and increment version number

the-real-adammork committed:
update lock files

the-real-adammork committed:
update comments

the-real-adammork committed:
update comments

carybakker commented on Sources/Ledger/PartialTxOut.swift:
```diff
@@ -57,4 +58,21 @@ extension PartialTxOut {
+        
+        let reconstructedCrc32 = TxOutUtils.calculateCrc32(from: reconstructed)
+        
+        return reconstructedCrc32 == sentCrc32 || reconstructedCrc32 == sentCommitmentCrc32
```
Why do we use || instad of && here?

the-real-adammork replied:
because the crc32 isn't always sent in the protobuf

if it is sent then we have to reconstruct the commitment data from the other fields in the TxOut (and our account keys) and then calculate the crc32 from the reconstructed commitment for comparison.

If crc32 is not sent (current + soon to be deprecated) then we still reconstruct the commitment data and then calculate crc32 for sent & reconstructed and compare.

Only one side can be true, which is sufficient.

carybakker committed:
Added convenience crc32 calculated variables for Data and Data32 and used to simplify code a bit

the-real-adammork committed:
Merge pull request #2 from carybakker/task/cary-crc32-updates

Added convenience crc32 calculated variables for Data and Data32

carybakker committed:
Minor code cleanup:

 - removed explicit init calls
 - removed unnecessary try
 - removed whitespace and decreased indentation for string to remove lint warnings

the-real-adammork committed:
Merge pull request #3 from carybakker/task/cary-crc32-updates-2

Minor code cleanup:

voloshyn reviewed (approved):
Checking for crc32 presence is a correct implementation, but for the historical context: the crc32 was not present for the transition period from a full tx_out_amount_commitment_data to reconstructed, but it is always sent back now.
https://github.com/mobilecoinofficial/android-sdk/blob/842f9194a736c24ed2bc7ed9c4e8acc1c09c2310/android-sdk/src/main/proto/view.proto#L260

the-real-adammork committed:
Re-organize some code, change logger.warning to assertionFailure, reduce verbosity of extensions and file names

the-real-adammork commented on Sources/Ledger/TxOutUtils.swift:
```diff
@@ -101,6 +99,35 @@ enum TxOutUtils {
+                case .invalidInput:
+                    // Safety: This condition indicates a programming error and can only
+                    // happen if arguments to mc_tx_out_commitment_crc32 are supplied incorrectly.
+                    logger.assertionFailure("error: \(redacting: error)")
```
@carybakker I changed these to assertionFailures which will act like "fatalError" during debug, but not in release.

carybakker replied:
ah, k... sounds good

the-real-adammork commented on Sources/Ledger/TxOutUtils.swift:
```diff
@@ -101,6 +99,35 @@ enum TxOutUtils {
         }
     }
 
+    static func calculateCrc32(
```
Moved this call back to TxOutUtils so it fits in better with other similar calls

carybakker replied:
@the-real-adammork I don't think anything is using this any longer - since you can just do <data>.commitmentCrc32 now, anyplace that was using this is now using the commitmentCrc32 ... or did I miss something?

carybakker replied:
scratch that - the impl is here, and Data32 calls it, instead of the impl being in Data32...

the-real-adammork commented on Sources/Utils/Data/Data+CommitmentCrc32.swift:
```diff
@@ -0,0 +1,14 @@
+    // this function is only valid in the following case:
+    //  - the Data instance is also a valid Data32
+    //  - the Data instance represents a CompressedCommitment
+    var commitmentCrc32: UInt32? {
```
made the extension name less verbose

carybakker reviewed (approved):
LGTM

the-real-adammork committed:
update podspec to require pre8 of libmobilecoin

the-real-adammork committed:
update locks

the-real-adammork committed:
Merge branch 'master' into task/correct-crc32-fix

the-real-adammork committed:
merge in latest from master

the-real-adammork commented:
@carybakker  Thank you for your feedback and changes, it made this PR much better.
